File: packages/sdk-maui/Platforms/iOS/Screeb.iOS.cs
#if IOS
using Foundation;
using Screeb.iOS.Binding;
using NativeScreeb = global::Screeb.iOS.Binding.Screeb;

namespace Screeb.Maui;

public static partial class Screeb
{
    public static partial Task<bool?> InitSdk(
        string channelId, string? userId, Dictionary<string, object>? properties,
        ScreebHooks? hooks, ScreebInitOptions? initOptions, string? language)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.SetSecondarySDK("maui", "0.1.0");
                var ioOpts = new InitOptions(NSDictionary.FromObjectsAndKeys(
                    new object[] { 
                        NSNumber.FromBoolean(initOptions?.IsDebugMode ?? false), 
                        NSNumber.FromBoolean(initOptions?.DisableMirror ?? false) 
                    },
                    new object[] { "isDebugMode", "disableMirror" }));
                var jHooks = HooksIOS.ToNSDictionary(HooksRegistry.RegisterHooks(hooks));
                NativeScreeb.InitSdk(
                    null, channelId, userId,
                    ToNSDictionary(ScreebUtils.FormatProperties(properties)) ?? new NSDictionary(),
                    ioOpts, jHooks, language);
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> CloseSdk()
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try { NativeScreeb.CloseSdk(); tcs.SetResult(true); }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> SetIdentity(string userId, Dictionary<string, object>? properties)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.SetIdentity(userId,
                    ToNSDictionary(ScreebUtils.FormatProperties(properties)) ?? new NSDictionary());
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> SetProperties(Dictionary<string, object>? properties)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.VisitorProperty(
                    ToNSDictionary(ScreebUtils.FormatProperties(properties)) ?? new NSDictionary());
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> ResetIdentity()
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try { NativeScreeb.ResetIdentity(); tcs.SetResult(true); }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<Dictionary<string, object>?> GetIdentity()
    {
        var tcs = new TaskCompletionSource<Dictionary<string, object>?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.GetIdentity((identity, error) =>
                {
                    if (error != null) tcs.SetException(new Exception(error.LocalizedDescription));
                    else tcs.SetResult(FromNSDictionary(identity));
                });
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> AssignGroup(string? groupType, string groupName, Dictionary<string, object>? properties)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.AssignGroup(groupType, groupName,
                    ToNSDictionary(ScreebUtils.FormatProperties(properties)) ?? new NSDictionary());
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> UnassignGroup(string? groupType, string groupName, Dictionary<string, object>? properties)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.UnassignGroup(groupType, groupName,
                    ToNSDictionary(ScreebUtils.FormatProperties(properties)) ?? new NSDictionary());
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> TrackEvent(string name, Dictionary<string, object>? properties)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.TrackEvent(name,
                    ToNSDictionary(ScreebUtils.FormatProperties(properties)) ?? new NSDictionary());
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> TrackScreen(string name, Dictionary<string, object>? properties)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.TrackScreen(name,
                    ToNSDictionary(ScreebUtils.FormatProperties(properties)) ?? new NSDictionary());
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> StartSurvey(
        string surveyId, bool allowMultipleResponses, Dictionary<string, object>? hiddenFields,
        bool ignoreSurveyStatus, ScreebHooks? hooks, string? language, string? distributionId)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                var jHooks = HooksIOS.ToNSDictionary(HooksRegistry.RegisterHooks(hooks));
                NativeScreeb.StartSurvey(
                    surveyId, allowMultipleResponses,
                    ToNSDictionary(ScreebUtils.FormatProperties(hiddenFields)) ?? new NSDictionary(),
                    ignoreSurveyStatus, jHooks, language, distributionId);
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> CloseSurvey(string? surveyId)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try { NativeScreeb.CloseSurvey(surveyId); tcs.SetResult(true); }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> StartMessage(
        string messageId, bool allowMultipleResponses, Dictionary<string, object>? hiddenFields,
        bool ignoreMessageStatus, ScreebHooks? hooks, string? language, string? distributionId)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                var jHooks = HooksIOS.ToNSDictionary(HooksRegistry.RegisterHooks(hooks));
                NativeScreeb.StartMessage(
                    messageId, allowMultipleResponses,
                    ToNSDictionary(ScreebUtils.FormatProperties(hiddenFields)) ?? new NSDictionary(),
                    ignoreMessageStatus, jHooks, language, distributionId);
                tcs.SetResult(true);
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> CloseMessage(string? messageId)
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try { NativeScreeb.CloseMessage(messageId); tcs.SetResult(true); }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> SessionReplayStart()
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try { NativeScreeb.SessionReplayStart(); tcs.SetResult(true); }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<bool?> SessionReplayStop()
    {
        var tcs = new TaskCompletionSource<bool?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try { NativeScreeb.SessionReplayStop(); tcs.SetResult(true); }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<string?> Debug()
    {
        var tcs = new TaskCompletionSource<string?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.Debug((info, error) =>
                {
                    if (error != null) tcs.SetException(new Exception(error.LocalizedDescription));
                    else tcs.SetResult(info);
                });
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    public static partial Task<string?> DebugTargeting()
    {
        var tcs = new TaskCompletionSource<string?>();
        NSRunLoop.Main.BeginInvokeOnMainThread(() =>
        {
            try
            {
                NativeScreeb.DebugTargeting((info, error) =>
                {
                    if (error != null) tcs.SetException(new Exception(error.LocalizedDescription));
                    else tcs.SetResult(info);
                });
            }
            catch (Exception ex) { tcs.SetException(ex); }
        });
        return tcs.Task;
    }

    // --- Helpers ---

    private static NSDictionary? ToNSDictionary(Dictionary<string, object>? dict)
    {
        if (dict == null) return null;
        var keys = new List<NSObject>();
        var values = new List<NSObject>();
        foreach (var (k, v) in dict)
        {
            keys.Add(new NSString(k));
            values.Add(v is string s ? new NSString(s) :
                       v is bool b ? NSNumber.FromBoolean(b) :
                       v is int i ? NSNumber.FromInt32(i) :
                       v is long l ? NSNumber.FromInt64(l) :
                       v is double d ? NSNumber.FromDouble(d) :
                       new NSString(v.ToString() ?? ""));
        }
        return NSDictionary.FromObjectsAndKeys(values.ToArray(), keys.ToArray());
    }

    private static Dictionary<string, object>? FromNSDictionary(NSDictionary? dict)
    {
        if (dict == null) return null;
        var result = new Dictionary<string, object>();
        foreach (var key in dict.Keys)
            result[key.ToString() ?? ""] = dict[key]?.ToString() ?? "";
        return result;
    }
}
#endif

File: packages/sdk-kmp/src/iosMain/kotlin/app/screeb/sdk/kmp/Screeb.ios.kt
@file:OptIn(kotlinx.cinterop.ExperimentalForeignApi::class)

package app.screeb.sdk.kmp

import app.screeb.sdk.ios.cinterop.InitOptions
import app.screeb.sdk.ios.cinterop.Screeb as NativeScreeb
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/** Converts a KMP property map to a Kotlin Map that the cinterop bridge converts to NSDictionary. */
@Suppress("UNCHECKED_CAST")
private fun Map<String, Any>?.toObjCMap(): Map<Any?, Any?> =
    (ScreebUtils.formatProperties(this) ?: emptyMap()) as Map<Any?, Any?>

/** Reads the cinterop-bridged NSDictionary (exposed as Map<Any?, *>) to a typed Kotlin Map. */
@Suppress("UNCHECKED_CAST")
private fun Map<Any?, *>?.toKotlinMap(): Map<String, Any>? {
    if (this == null) return null
    return entries.associate { (k, v) -> k.toString() to (v?.toString() ?: "") }
}

actual object Screeb {

    actual suspend fun initSdk(
        channelId: String,
        userId: String?,
        properties: Map<String, Any>?,
        hooks: ScreebHooks?,
        initOptions: ScreebInitOptions?,
        language: String?,
    ): Boolean? = withContext(Dispatchers.Main) {
        runCatching {
            val uuidMap = HooksRegistry.register(hooks)
            NativeScreeb.setSecondarySDK("kmp", version = SDK_VERSION)
            val opts = InitOptions()
            opts.setIsDebugMode(initOptions?.isDebugMode ?: false)
            opts.setDisableMirror(initOptions?.disableMirror ?: false)
            NativeScreeb.initSdk(
                null,
                channelId = channelId,
                identity = userId,
                visitorProperty = properties.toObjCMap(),
                initOptions = opts,
                hooks = HooksIOS.toMap(uuidMap),
                language = language,
            )
            true
        }.orNullRethrowingCancellation()
    }

    actual suspend fun closeSdk(): Boolean? = withContext(Dispatchers.Main) {
        runCatching {
            HooksRegistry.clear()
            HooksScope.reset()
            NativeScreeb.closeSdk()
            true
        }.orNullRethrowingCancellation()
    }

    actual suspend fun setIdentity(
        userId: String,
        properties: Map<String, Any>?,
    ): Boolean? = withContext(Dispatchers.Main) {
        runCatching {
            NativeScreeb.setIdentity(
                userId,
                visitorProperty = properties.toObjCMap(),
            )
            true
        }.orNullRethrowingCancellation()
    }

    actual suspend fun setProperties(properties: Map<String, Any>?): Boolean? =
        withContext(Dispatchers.Main) {
            runCatching {
                NativeScreeb.visitorProperty(properties.toObjCMap())
                true
            }.orNullRethrowingCancellation()
        }

    actual suspend fun resetIdentity(): Boolean? = withContext(Dispatchers.Main) {
        runCatching { NativeScreeb.resetIdentity(); true }.orNullRethrowingCancellation()
    }

    actual suspend fun getIdentity(): Map<String, Any>? = withContext(Dispatchers.Main) {
        runCatching {
            suspendCancellableCoroutine { cont ->
                NativeScreeb.getIdentity { identity, error ->
                    if (error != null) cont.resumeWithException(Exception(error.localizedDescription))
                    else cont.resume((identity as Map<Any?, *>?).toKotlinMap())
                }
            }
        }.orNullRethrowingCancellation()
    }

    actual suspend fun assignGroup(
        groupType: String?,
        groupName: String,
        properties: Map<String, Any>?,
    ): Boolean? = withContext(Dispatchers.Main) {
        runCatching {
            NativeScreeb.assignGroup(
                groupType,
                name = groupName,
                properties = properties.toObjCMap(),
            )
            true
        }.orNullRethrowingCancellation()
    }

    actual suspend fun unassignGroup(
        groupType: String?,
        groupName: String,
        properties: Map<String, Any>?,
    ): Boolean? = withContext(Dispatchers.Main) {
        runCatching {
            NativeScreeb.unassignGroup(
                groupType,
                name = groupName,
                properties = properties.toObjCMap(),
            )
            true
        }.orNullRethrowingCancellation()
    }

    actual suspend fun trackEvent(name: String, properties: Map<String, Any>?): Boolean? =
        withContext(Dispatchers.Main) {
            runCatching {
                NativeScreeb.trackEvent(
                    name,
                    trackingEventProperties = properties.toObjCMap(),
                )
                true
            }.orNullRethrowingCancellation()
        }

    actual suspend fun trackScreen(name: String, properties: Map<String, Any>?): Boolean? =
        withContext(Dispatchers.Main) {
            runCatching {
                NativeScreeb.trackScreen(
                    name,
                    trackingEventProperties = properties.toObjCMap(),
                )
                true
            }.orNullRethrowingCancellation()
        }

    actual suspend fun startSurvey(
        surveyId: String,
        allowMultipleResponses: Boolean,
        hiddenFields: Map<String, Any>?,
        ignoreSurveyStatus: Boolean,
        hooks: ScreebHooks?,
        language: String?,
        distributionId: String?,
    ): Boolean? = withContext(Dispatchers.Main) {
        runCatching {
            val uuidMap = HooksRegistry.register(hooks)
            NativeScreeb.startSurvey(
                surveyId,
                allowMultipleResponses = allowMultipleResponses,
                hiddenFields = hiddenFields.toObjCMap(),
                ignoreSurveyStatus = ignoreSurveyStatus,
                hooks = HooksIOS.toMap(uuidMap),
                language = language,
                distributionId = distributionId,
            )
            true
        }.orNullRethrowingCancellation()
    }

    actual suspend fun closeSurvey(surveyId: String?): Boolean? = withContext(Dispatchers.Main) {
        runCatching { NativeScreeb.closeSurvey(surveyId); true }.orNullRethrowingCancellation()
    }

    actual suspend fun startMessage(
        messageId: String,
        allowMultipleResponses: Boolean,
        hiddenFields: Map<String, Any>?,
        ignoreMessageStatus: Boolean,
        hooks: ScreebHooks?,
        language: String?,
        distributionId: String?,
    ): Boolean? = withContext(Dispatchers.Main) {
        runCatching {
            val uuidMap = HooksRegistry.register(hooks)
            NativeScreeb.startMessage(
                messageId,
                allowMultipleResponses = allowMultipleResponses,
                hiddenFields = hiddenFields.toObjCMap(),
                ignoreMessageStatus = ignoreMessageStatus,
                hooks = HooksIOS.toMap(uuidMap),
                language = language,
                distributionId = distributionId,
            )
            true
        }.orNullRethrowingCancellation()
    }

    actual suspend fun closeMessage(messageId: String?): Boolean? = withContext(Dispatchers.Main) {
        runCatching { NativeScreeb.closeMessage(messageId); true }.orNullRethrowingCancellation()
    }

    actual suspend fun sessionReplayStart(): Boolean? = withContext(Dispatchers.Main) {
        runCatching { NativeScreeb.sessionReplayStart(); true }.orNullRethrowingCancellation()
    }

    actual suspend fun sessionReplayStop(): Boolean? = withContext(Dispatchers.Main) {
        runCatching { NativeScreeb.sessionReplayStop(); true }.orNullRethrowingCancellation()
    }

    actual suspend fun debug(): String? = withContext(Dispatchers.Main) {
        runCatching {
            suspendCancellableCoroutine { cont ->
                NativeScreeb.debug { info, error ->
                    if (error != null) cont.resumeWithException(Exception(error.localizedDescription))
                    else cont.resume(info)
                }
            }
        }.orNullRethrowingCancellation()
    }

    actual suspend fun debugTargeting(): String? = withContext(Dispatchers.Main) {
        runCatching {
            suspendCancellableCoroutine { cont ->
                NativeScreeb.debugTargeting { info, error ->
                    if (error != null) cont.resumeWithException(Exception(error.localizedDescription))
                    else cont.resume(info)
                }
            }
        }.orNullRethrowingCancellation()
    }

    // Forward to the native SDK (editor/survey/message links open in-app). Call it
    // from your AppDelegate's open-URL entry point or a shared linking listener.
    actual fun handleDeepLink(url: String?) {
        if (url == null) return
        NativeScreeb.handleDeepLink(platform.Foundation.NSURL.URLWithString(url))
    }
}
